### ISSUE TITLE
Ban GradScaler scale from being less than 1

### ISSUE THREAD

**Issue by janeyx99** [pytorch/pytorch#99640]
### 🚀 The feature, motivation and pitch

We should standardize on a way to tell users when a reasonable/good scale for the AMP GradScaler cannot be found.

Context:

The grad scaler is awesome in that it "automatically" finetunes itself to settle on an ideal scale. However! There are instances where there is just no ideal scale for certain parameters, for example the one mentioned in issue #96755, where the scale becomes so so tiny that it causes infs and, consequently, nans. In this issue, we were nestled in the unlikely (but clearly not impossible) world where for a certain scale S, the scaled gradients were valid, but the unscaled (g/S) were invalid, leading to weird inconsistent behavior.

Generally, we want to be add invariants/guarantees to prevent weird inconsistent behavior and give users faster and clearer signal. Since the purpose of the scale is to scale UP the loss so that the grads do not underflow in fp16, we should set a hard boundary of scale must be greater than or equal to 1. Note that adding a check may induce a slight perf hit.

cc @ngimel @crcrpar @stas00 @albanD

### Alternatives

Do nothing, let users deal with nans from the perspective of BC + that there might be currently unimagined worlds where users want to handle nans themselves (?)

### Additional context

More context provided in https://github.com/pytorch/pytorch/issues/96755.

**Comment by albanD** [comment]
Sounds like a good spot for a warning?
Or we could deprecate/remove it and allow this unsafe behavior by passing an explicit flag?

**Comment by ptrblck** [comment]
> There are instances where there is just no ideal scale for certain parameters, for example the one mentioned in issue https://github.com/pytorch/pytorch/issues/96755, where the scale becomes so so tiny that it causes infs and, consequently, nans.

I'm wondering if this example is a good reference since in the issue you also explained:
> based on @crcrpar's benchmarking results in https://github.com/pytorch/pytorch/pull/98620#issuecomment-1512227931, it looks like the "better" loss would be a nan under the hood without the logging_nan_inf_filter argument.

So does it mean the reference training is indeed not stable, overflows in the forward pass, hides the NaNs/Infs from the user, and thus drops the gradient scaling factor massively? Note that overflows in the *forward* pass are not expected and indicate a broken training.

> Generally, we want to be add invariants/guarantees to prevent weird inconsistent behavior and give users faster and clearer signal. 

I totally agree that a clear signal is fantastic. However, based on my understanding it feels as if the clear signal was already hidden from the user and we did not see that the forward pass is already overflowing (please correct me if I'm misinterpreting the linked issue). If so, I'm unsure what exactly we are fixing with this RFE besides potentially hiding more issues from the user.

> Or we could deprecate/remove it and allow this unsafe behavior by passing an explicit flag?

This would be desirable if this additional check adds a synchronization thus not only creating a perf. hit but also disallowing CUDA Graphs use cases.

**Comment by stas00** [comment]
Actually that's a good point - as t5 models are bf16-pretrained - most of them [overflow when finetuned under fp16 regime](https://github.com/huggingface/transformers/pull/10956). I think I have only seen `t5-small` being ok, anything larger than it overflows.

I probably should have picked a different type of model for the benchmark in the first place! I completely forgot about this issue.

But on the other hand this is a good "bad" example where the training should have failed and it didn't!

**Comment by janeyx99** [comment]
@ptrblck 
> I totally agree that a clear signal is fantastic. However, based on my understanding it feels as if the clear signal was already hidden from the user and we did not see that the forward pass is already overflowing (please correct me if I'm misinterpreting the linked issue). If so, I'm unsure what exactly we are fixing with this RFE besides potentially hiding more issues from the user.

I think there are two obfuscations to detangle. The first is that the linked example was hid NaNs from users, which I'll leave up to @stas00 to report/fix/do with what hf will because this isn't something we can change from the PyTorch end. The second is that we allow the grad scale, which is meant to scale things up, to fall under 1. For the linked issue, having this check would be a faster signal, though I see your point that it isn't necessarily clearer since the training may have been borked long ago. The question that I have is then--would this check still be useful for other cases?

100% agreed with the idea that we should not incur a perf hit/prevent CUDA graphs by default without a strong enough reason to. I see three options going forward (not all mutually exclusive):
1. We don't change the code, but we document that the grad scale has no guarantees and COULD go smaller than 1. Thus, when users run into unexpected NaNs, there is an explanation somewhere for them + they can debug from there.
2. We could add a flag that would do the check. This is the backwards compatible way, where users can enroll into the check to verify their models are functioning okay. Once verified, one can remove the flag to opt into better perf.
3. If more people report scale becoming small being a problem for them (so if there _are_ other use cases that would find this check helpful), then we can flip the default in option 2 and have the flag opt out of the check.

It would be good to know how often an invalid scale is actually a problem--if in most cases training was already broken in the forward, then this would not be worth adding.

**Comment by crcrpar** [comment]
> 1. We don't change the code, but we document that the grad scale has no guarantees and COULD go smaller than 1. Thus, when users run into unexpected NaNs, there is an explanation somewhere for them + they can debug from there.

I find this option appropriate at least for now since the check can be written in call sites w/o creating a custom C++/CUDA extension. Also, in that way, it'd be clearer that the check would sacrifice perf and/or cuda graphs.

some questions:
- unexpected Nans of what? loss, parameters, and/or grads?
- what kind of information would you expect to help users debug?

**Comment by stas00** [comment]
> The first is that the linked example was hid NaNs from users, which I'll leave up to @stas00 to report/fix/do with what hf will because this isn't something we can change from the PyTorch end. 

To clarify - there is nothing to fix on the HF side. bf16-pretrained models most of the time can't operate in the fp16 numerical range of max 64k. I linked to the PR that wasn't merged since it was an attempt at a workaround, but it's not great as it does too much upcasting to fp32 defeating the purpose of AMP. The problem is the weights, not the modeling code.

So I should give you an fp16-pretrained model to do comparisons with. Let me find a suitable one and I will share then.

**Comment by janeyx99** [comment]
>unexpected Nans of what? loss, parameters, and/or grads?
In this particular example, it would be grads which would propagate to parameters (and eventually loss), no?

> what kind of information would you expect to help users debug?
in my head, I'm envisioning an error message for when someone runs their model with fp16 that says "Scale is about to go below 1! There is likely no valid scale for which fp16 works with your current weights."

**Comment by janeyx99** [comment]
Okay, I am inclined to start by doing the first thing (documenting) as that should only help. 

I'm not yet convinced that adding a check would be worth the potential perf cost and eventual maintenance cost for the seemingly small number of cases that will hit this edge case, but I will leave this issue open to allow potential future comments to change my mind.

**Comment by janeyx99** [comment]
And to relate @albanD's tag in the DebugMode RFC, steps 2 and 3 from above could be integrated with a more global debug mode that users could turn on to get more debug messages and checks at the cost of perf.
